### PR TITLE
[quantization] Add `calibrate_seq_len`

### PR DESCRIPTION
**Pull request by stamalakhov**
This PR adds `calibrate_seq_len` to get better results on accuracy.

please see results  [here](https://github.com/Samsung/TICO/pull/533#issuecomment-3990447438)

<details>
<summary>log of `python tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py --model "HuggingFaceTB/SmolLM2-135M-Instruct"  --max_seq_len 256 --gptq_mse "--eval_tasks" "winogrande,arc_easy,arc_challenge,openbookqa" ...`</summary>

```

Namespace(model='HuggingFaceTB/SmolLM2-135M-Instruct', device='cuda', dtype='float32', seed=42, trust_remote_code=False, hf_token=None, no_tqdm=False, no_GPTQ=False, no_PTQ=False, save_circle_to_folder='/home/stanislav', cache_dir='/mnt/storage/transformers_cache', nsamples_for_qcalibration=128, linear_weight_bits=4, gptq_mse=True, max_seq_len=256, calibrate_seq_len=2048, embedding_weight_bits=8, lm_head_weight_bits=4, eval_tasks='winogrande,arc_easy,arc_challenge,openbookqa')
=== Config ===
Model            : HuggingFaceTB/SmolLM2-135M-Instruct
Device           : cuda
DType            : float32

Loading FP model …
`torch_dtype` is deprecated! Use `dtype` instead!
Loading weights: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:01<00:00, 170.72it/s, Materializing param=model.norm.weight]

Calculating original perplexities …
Token indices sequence length is longer than the specified maximum sequence length for this model (304978 > 8192). Running this sequence through the model will result in indexing errors
PPL: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊| 1191/1192 [01:38<00:00, 12.15it/s]

┌── Wikitext-2 test perplexity ─────────────
│ FP32 :    30.32
└───────────────────────────────────────────
`pretrained` model kwarg is not of type `str`. Many other model arguments may be ignored. Please do not launch via accelerate or use `parallelize=True` if passing an existing model this way.
Passed an already-initialized model through `pretrained`, assuming single-process call to evaluate() or custom distributed integration
Warning: You are sending unauthenticated requests to the HF Hub. Please set a HF_TOKEN to enable higher rate limits and faster downloads.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 500/500 [00:03<00:00, 141.97it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1172/1172 [00:12<00:00, 96.22it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2376/2376 [00:21<00:00, 108.32it/s]
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1267/1267 [00:00<00:00, 11541.23it/s]
Running loglikelihood requests: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 18722/18722 [21:08<00:00, 14.76it/s]
Original RESULTS ARE:
|    Tasks    |Version|Filter|n-shot| Metric |   |Value |   |Stderr|
|-------------|------:|------|-----:|--------|---|-----:|---|-----:|
|arc_challenge|      1|none  |     0|acc     |↑  |0.2594|±  |0.0128|
|             |       |none  |     0|acc_norm|↑  |0.2773|±  |0.0131|
|arc_easy     |      1|none  |     0|acc     |↑  |0.5400|±  |0.0102|
|             |       |none  |     0|acc_norm|↑  |0.4882|±  |0.0103|
|openbookqa   |      1|none  |     0|acc     |↑  |0.2240|±  |0.0187|
|             |       |none  |     0|acc_norm|↑  |0.3320|±  |0.0211|
|winogrande   |      1|none  |     0|acc     |↑  |0.5107|±  |0.0140|

Applying GPTQ …
Quantizing layers: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [02:50<00:00,  5.67s/layer]
Wrapping layers with PTQWrapper …                                                                                                                                                                                                                         
Calibrating PTQ obeservers…
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [02:24<00:00,  1.13s/it]

Calculating perplexities …
PPL: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊| 1191/1192 [10:56<00:00,  1.81it/s]

┌── Wikitext-2 test perplexity ─────────────
│ int16 :    52.08
└───────────────────────────────────────────
`pretrained` model kwarg is not of type `str`. Many other model arguments may be ignored. Please do not launch via accelerate or use `parallelize=True` if passing an existing model this way.
Passed an already-initialized model through `pretrained`, assuming single-process call to evaluate() or custom distributed integration
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 500/500 [00:05<00:00, 87.04it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1172/1172 [00:12<00:00, 91.11it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2376/2376 [00:24<00:00, 96.53it/s]
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1267/1267 [00:00<00:00, 9694.34it/s]
Running loglikelihood requests: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 18722/18722 [1:29:06<00:00,  3.50it/s]
Quantized RESULTS ARE:
|    Tasks    |Version|Filter|n-shot| Metric |   |Value |   |Stderr|
|-------------|------:|------|-----:|--------|---|-----:|---|-----:|
|arc_challenge|      1|none  |     0|acc     |↑  |0.2560|±  |0.0128|
|             |       |none  |     0|acc_norm|↑  |0.2790|±  |0.0131|
|arc_easy     |      1|none  |     0|acc     |↑  |0.4617|±  |0.0102|
|             |       |none  |     0|acc_norm|↑  |0.4470|±  |0.0102|
|openbookqa   |      1|none  |     0|acc     |↑  |0.1960|±  |0.0178|
|             |       |none  |     0|acc_norm|↑  |0.2980|±  |0.0205|
|winogrande   |      1|none  |     0|acc     |↑  |0.5328|±  |0.0140|

saving the whole model to /home/stanislav/model.q.circle

```

</details>

TICO-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

**Comment by stamalakhov** [comment]
> > This PR adds calibrate_seq_len to get better results on accuracy.
> 
> Could you elaborate more about this change?
> 
> 1. What this `calibrate_seq_len` is for?

@mhs4670go 
It occured that the more `seq_len` was used for calibration, we will get the better accuracy. This is important for `GPTQ`. 

> 2. Why does it make better results?

IMHO
Seems like calibration with bigger `seq_len` produces more `outliers` (especially in inner layers). The calibration set is more challenging, so quantized model performs better afterwards.

Just an example with [`unsloth/Llama-3.2-3B-Instruct`](https://huggingface.co/unsloth/Llama-3.2-3B-Instruct),  `seq_len == 256`:

| Config ID | PPL  |arc_easy(%)|arc_challenge (%)|winogrande (%)|openbookqa(%)|
|-----------|------|--|-|-|-|
| FP32 | 19.05 |75|44|67|29|
| GPTQ_MSE_w4A16| 22.9|72| 38|64|26|
| GPTQ_MSE_w4A16_#526_equalization| **22.8**|**72**| **41**|66|27|
| GPTQ_MSE_w4A16_#533_equalization| **22.8**|**72**| **41**|**67**|**28**|

So seems like this PR produces better results.

**Comment by stamalakhov** [comment]
Comparison of `main` branch accuracy with this PR.

For [`unsloth/Llama-3.2-3B-Instruct`](https://huggingface.co/unsloth/Llama-3.2-3B-Instruct),  `seq_len == 256` and different `calibrate_seq_len` :

| Config ID | PPL  |arc_easy(%)|arc_challenge (%)|winogrande (%)|openbookqa(%)|
|-----------|------|--|-|-|-|
| FP32 | 19.05 |75|44|67|29|
| GPTQ_MSE_w4A16_main| 22.9|72| 38|64|26|
| GPTQ_MSE_w4A16_#533_equalization_with_2048_calibrate_seq_len| **22.8**|**72**| 41|**67**|**28**|
| GPTQ_MSE_w4A16_#533_equalization_with_4096_calibrate_seq_len| 22.9|71| **42**|**67**|27|


TODO: other models

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 1]
```diff
@@ -327,10 +349,16 @@ def main():
         help="Whether to use mse in gptq",
     )
     parser.add_argument(
-        "--max_seq_len",
+        "--convert_seq_len",
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: Ok.

File: tico/quantization/wrapq/wrappers/llama/quant_model.py [hunk 2]
```diff
@@ -209,8 +209,8 @@ def forward(
         position_embeddings = self.get_position_embeddings_for(hidden_states)
         cos, sin = position_embeddings
         position_embeddings = (
-            self._fq(cos, self.obs_cos),
-            self._fq(sin, self.obs_sin),
+            self._fq(cos[:, : hidden_states.size(1), :], self.obs_cos),
+            self._fq(sin[:, : hidden_states.size(1), :], self.obs_sin),
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_model.py]: @mhs4670go 
Sorry for a lack of details.
`attention_mask` is prepared for current `seq_len` as 
```
return self.causal_mask_template[..., :seq_len, :seq_len].to(device)
```
so in case related `causal_mask_template` size is larger then `seq_len` everything is fine (Because it's just upper matrix filled with constant).
We can do the same here:
prepare `(rope_cos_template, rope_sin_template)` for larger `seq_len` and then just extract what is needed for current `seq_len`.
It is assumed that `calibrate_seq_len >= max_seq_len`.

[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_model.py]: Yep.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 3]
```diff
@@ -233,6 +244,15 @@ def quantize_using_PTQ(q_m, calib_inputs, args):
...
 
 def evaluate(q_m, tokenizer, dataset_test, args):
+
+    if hasattr(q_m, "wrapped"):
+        config = q_m.wrapped.config
+    else:
+        config = q_m.config
+
+    orig_seq_len = config.max_position_embeddings
+    config.max_position_embeddings = args.max_seq_len
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: @mhs4670go
Done. At least ppl is fine for `SmolLM`. I'll check for `eval_tasks` (its around 1.5 hour)
